core/src/main/java/org/apache/calcite/rel/metadata/RelMdForeignKeys.java

```diff
@@ -174,72 +176,88 @@ private static ImmutableBitSet getProjectForeignKeys(SingleRel rel, RelMetadataQ
         allOutColumns = allOutColumns.union(outColumns);
       }
     }
-    return allOutColumns.isEmpty() ? EMPTY_BIT_SET : allOutColumns;
+    return allOutColumns;
   }
 
   public ImmutableBitSet getForeignKeys(TableScan rel, RelMetadataQuery mq,
-      boolean ignoreNulls) {
+      boolean containNulls) {
     final RelOptTable table = rel.getTable();
     final BuiltInMetadata.ForeignKeys.Handler handler =
         table.unwrap(BuiltInMetadata.ForeignKeys.Handler.class);
     if (handler != null) {
-      return handler.getForeignKeys(rel, mq, ignoreNulls);
+      return handler.getForeignKeys(rel, mq, containNulls);
     }
 
     final List<RelReferentialConstraint> referentialConstraints =
         table.getReferentialConstraints();
-    if (referentialConstraints == null || referentialConstraints.size() == 0) {
+    if (referentialConstraints == null || referentialConstraints.isEmpty()) {
       return EMPTY_BIT_SET;
     }
-    final List<IntPair> foreignUniquePair = referentialConstraints.stream()
+    Set<Integer> foreignKeys = referentialConstraints.stream()
         .map(RelReferentialConstraint::getColumnPairs)
         .flatMap(Collection::stream)
-        .collect(Collectors.toList());
+        .map(pair -> {
+          return pair.source;
+        })
+        .collect(Collectors.toSet());
 
-    List<Integer> foreignKeys = IntPair.left(foreignUniquePair);
-    if (!ignoreNulls) {
+    if (!containNulls) {
       final List<RelDataTypeField> fieldList = rel.getRowType().getFieldList();
       foreignKeys = foreignKeys.stream()
           .filter(index -> !fieldList.get(index).getType().isNullable())
-          .collect(Collectors.toList());
+          .collect(Collectors.toSet());
     }
-    return foreignKeys.isEmpty() ? EMPTY_BIT_SET : ImmutableBitSet.of(foreignKeys);
+    return ImmutableBitSet.of(foreignKeys);
   }
 
-  public ImmutableBitSet getForeignKeys(SetOp rel, RelMetadataQuery mq,
-      boolean ignoreNulls) {
+  /**
+   * The foreign keys of Union are precisely the intersection of its every
+   * input foreign keys.
+   */
+  public ImmutableBitSet getForeignKeys(Union rel, RelMetadataQuery mq,
+      boolean containNulls) {
 
     ImmutableBitSet foreignKeys = ImmutableBitSet.of();
     for (RelNode input : rel.getInputs()) {
-      ImmutableBitSet inputForeignKeys = mq.getForeignKeys(input, ignoreNulls);
+      ImmutableBitSet inputForeignKeys = mq.getForeignKeys(input, containNulls);
       if (inputForeignKeys.isEmpty()) {
         return EMPTY_BIT_SET;
       }
       foreignKeys = foreignKeys.isEmpty()
           ? inputForeignKeys : foreignKeys.intersect(inputForeignKeys);
     }
-    if (foreignKeys.isEmpty()) {
-      return EMPTY_BIT_SET;
+    return foreignKeys;
+  }
+
+  /**
+   * The foreign keys of Intersect are precisely the union set of its every
```

Comment:
I think it's the same as per `Union`, that is, the intersection of the FKs from the operands. Because you can apply this set operations on whatever pair of relations having the "same" schema type, but you can very well mix up columns with different FK relationship for the same position.

The same applies to `Minus`, so you probably want to revert back to a single implementation for `SetOp` and revise the tests.

In any case, `areForeignKeysValid` will bail out on such cases, as it will try to track down the original column via `RelMetadataQuery.getColumnOrigin()`, and if a mix is found, more origins will be detected, and it will return `null`.

Since it's a bit tricky I suggest to add at least some basic tests to `RelOptRulesTest` per each set operator to confirm our understanding (simple case and tricky one like a mix of column origin under the same position).

Reply:
OK, I get it.